toc/rfc/rfc-draft-bindable-secrets.md

```diff
@@ -0,0 +1,254 @@
+```
+
+If `SECRET_TYPE` is `value` then contents of `-p` will not be parsed. If `SECRET_TYPE` is `json`, `certificate`, or some other CredHub credential type, then the contents of `-p` MUST be valid JSON and may be parsed.
+
+This command will ultimately create a `POST` request to `/v3/secrets`. Example `curl` invocation:
+
+```console
+curl "https://api.example.org/v3/secrets" \
+  -X POST \
+  -H "Authorization: bearer [token]" \
+  -H "Content-type: application/json" \
+  -d '{
+    "type": "json",
+    "name": "my-app-secrets",
+    "value": {
+      "application.yml" : "---\napplication:\n ...",
+      "secret.txt": "my-secret"
+    }
+    "metadata": {
+      "annotations": {},
+      "labels": {}
+    },
+    "relationships": {
+      "space": {
+        "data": {
+          "guid": "7304bc3c-7010-11ea-8840-48bf6bec2d78"
+        }
+      }
+    }
+  }'
+```
+
+### Reading a `secret` and Its Value
+Users that have permission to view environment variables for an app will be able to view `secret` values within a `space`. We will introduce a separate endpoint for retrieving these (similar to how app environment variables and service instance credentials are retrieved): `GET /v3/secrets/:guid/value`.
+
+```console
+curl "https://api.example.org/v3/secrets/[guid]/value" \
+  -X GET \
+  -H "Authorization: bearer [token]"
+  
+"value" : {
+  "some-json-key" : "some-json-value"
+}
+```
+
+To support this, Cloud Controller will need to have both read and write permissions for the `secret`'s corresponding credential in CredHub.
+
+### Creating a `secret_binding`
+To create `secret_bindings`, users will use a new CLI command: `cf bind-secret`. This will create a `POST` request to `/v3/secret_bindings`.
+
+```console
+cf bind-secret APP_NAME SECRET_NAME BINDING_NAME
+```
+
+### CLI Changes
+Work will need to be done to implement the following commands:
+- `cf create-secret`
+- `cf bind-secret`
+- `cf secrets`
+- `cf secret`
+- `cf delete-secret`
+- `cf unbind-secret`
+
+In addition, modifications may be made to existing CLI commands to surface `secrets` and `secret_bindings`. For example, adding bound `secret`s to `cf app` output.
+
+### Cloud Controller Changes
+New database tables will need to be created to store `secret`s and `secret_binding`s. CRUD APIs will need to be created for these resources. We will also need an API to support viewing a `secret`'s value.
+
+### BBS changes
+BBS was recently updated to support `ServiceBindingFiles` as part of RFC 0030. A `secret`'s value may result in one or more files being created in the `tmpfs` mounted directory defined by the `MountPath` parameter.
+
+```
+action := &models.RunAction{
+  Path: "/path/to/executable",
+  Args: []string{"some", "args to", "pass in"},
+  Dir: "/path/to/working/directory",
+  User: "username",
+  EnvironmentVariables: []*models.EnvironmentVariable{
+    {
+      Name: "ENVNAME",
+      Value: "ENVVALUE",
+    },
+  },
+ ServiceBindingFiles: []*models.Files{
```

Comment:
Yeah, I agree.

@beyhan do you have thoughts on how to make this cleaner? How far along is the Service Binding Files work for us to potentially generalize the BBS interface?

Reply:
I think my comment https://github.com/cloudfoundry/community/pull/994#discussion_r1801149230 is also relevant here.

Reply:
I know this may be too late, but is the name of the field `ServiceBindingFiles` set in stone? I'm wondering if we could make it more generic -- even `VolumeMountedFiles` or something. I just don't know how far along the file-based service binding implementation is at this point.

cc @Gerg

Reply:
Ideally I would have noticed this in the file-based binding RFC, but this does feel like CF-level concepts leaking into BBS's domain model. Historically, BBS has tried to separate itself from CF concepts (e.g. `app` vs `LRP`).

Reply:
> Ideally I would have noticed this in the file-based binding RFC, but this does feel like CF-level concepts leaking into BBS's domain model. Historically, BBS has tried to separate itself from CF concepts (e.g. `app` vs `LRP`).

I think at the stage which we are with the implementation it should be possible to change `ServiceBindingFiles` to `VolumeMountedFiles`. @stephanme, @PlamenDoychev any thoughts on this?

Reply:
FYI: https://github.com/cloudfoundry/community/pull/1008

Reply:
Thanks @beyhan , I will update this RFC to reflect those changes.